packages/assets-controllers/src/CurrencyRateController.ts

```diff
@@ -163,69 +126,33 @@ export class CurrencyRateController extends BaseControllerV2<
    * @param currentCurrency - ISO 4217 currency code.
    */
   async setCurrentCurrency(currentCurrency: string) {
-    this.update((state) => {
-      state.pendingCurrentCurrency = currentCurrency;
-    });
-    await this.updateExchangeRate();
-  }
-
-  /**
-   * Sets a new native currency.
-   *
-   * @param symbol - Symbol for the base asset.
-   */
-  async setNativeCurrency(symbol: string) {
-    this.update((state) => {
-      state.pendingNativeCurrency = symbol;
-    });
-    await this.updateExchangeRate();
-  }
-
-  private stopPolling() {
-    if (this.intervalId) {
-      clearInterval(this.intervalId);
+    const releaseLock = await this.mutex.acquire();
+    const nativeCurrencies = Object.keys(this.state.currencyRates);
+    try {
+      this.update(() => {
+        return {
+          ...defaultState,
+          currentCurrency,
+        };
+      });
+    } finally {
+      releaseLock();
     }
+    nativeCurrencies.forEach(this.updateExchangeRate.bind(this));
   }
 
   /**
-   * Starts a new polling interval.
-   */
-  private async startPolling(): Promise<void> {
-    this.stopPolling();
-    // TODO: Expose polling currency rate update errors
-
-    await safelyExecute(async () => await this.updateExchangeRate());
-
-    this.intervalId = setInterval(async () => {
-      await safelyExecute(async () => await this.updateExchangeRate());
-    }, this.intervalDelay);
-  }
-
-  /**
-   * Updates exchange rate for the current currency.
+   * Updates the exchange rate for the current currency and native currency pair.
    *
-   * @returns The controller state.
+   * @param nativeCurrency - The ticker symbol for the chain.
    */
-  async updateExchangeRate(): Promise<CurrencyRateState | void> {
-    if (!this.#enabled) {
-      console.info(
-        '[CurrencyRateController] Not updating exchange rate since network requests have been disabled',
-      );
-      return this.state;
-    }
+  async updateExchangeRate(nativeCurrency: string): Promise<void> {
```

Comment:
would it be even safer to always pass the currentCurrency as an arg here as well... in case there is a race condition with the state update here: https://github.com/MetaMask/core/pull/1805/files#diff-9c8bbd91f0dff6d7000e65707a7045d650c9eaaa82984c2ff911ef559870a19bR132 ?

Reply:
modification and reads of state are both protected by mutexes in both this method and the `setCurrentCurrency` method you linked. There is no race condition afaik. The question is if the amount of code this mutex locks can be reduced or not.

In the case that there is already an `updateExchangeRate` running for a soon to be replaced currentCurrency (fiat, usually), it must finish running, update state, and release the lock it's holding before `setCurrentCurrency` is able to update the state for the new currentCurrency. Then any future updateExchangeRate executions will read and use the new currentCurrency from state